File: nosql-create-index-typescript/src/control-plane.ts
/**
 * Control-plane operations using @azure/arm-cosmosdb (ARM SDK).
 *
 *   1. Create containers with vector indexes
 *   2. Clean up sample-created containers
 *
 * RBAC Setup:
 * - Role definitions and assignments are created by `azd up` via Bicep (infra/database.bicep lines 195-226)
 * - Sample code uses DefaultAzureCredential() for authentication
 * - No RBAC creation code is needed in the sample
 */

import { CosmosDBManagementClient } from "@azure/arm-cosmosdb";
import type { TokenCredential } from "@azure/identity";
import type { SampleConfig } from "./config.js";

/** Create an ARM management client for Cosmos DB. */
export function createArmClient(
  credential: TokenCredential,
  subscriptionId: string
) {
  return new CosmosDBManagementClient(credential, subscriptionId);
}

// ---------------------------------------------------------------------------
// Delete container (if it exists) to ensure clean state
// ---------------------------------------------------------------------------
async function deleteContainerIfExists(
  armClient: CosmosDBManagementClient,
  config: SampleConfig,
  containerName?: string
) {
  const actualContainerName = containerName || config.cosmos.containerName;
  try {
    console.log(`  Deleting existing container if present...`);
    await armClient.sqlResources.beginDeleteSqlContainerAndWait(
      config.azure.resourceGroup!,
      config.cosmos.accountName!,
      config.cosmos.databaseName,
      actualContainerName
    );
    console.log(`  Deleted existing container`);
  } catch (error) {
    // 404 means container doesn't exist — that's fine
    if ((error as any).code === 404 || (error as any).message?.includes("NotFound")) {
      console.log(`  Container does not exist (OK)`);
    } else {
      throw error;
    }
  }
}

// ---------------------------------------------------------------------------
// Step 1 — Create container with vector index
// ---------------------------------------------------------------------------
export async function createContainer(
  armClient: CosmosDBManagementClient,
  config: SampleConfig
) {
  const indexTypes = [
    { type: "diskANN", containerName: config.cosmos.diskannContainerName },
    { type: "quantizedFlat", containerName: config.cosmos.quantizedflatContainerName },
  ];

  const embeddingPath = `/${config.embeddingField}`;

  for (const indexConfig of indexTypes) {
    console.log("\n=== Step 1: Create Container with Vector Index ===");
    console.log(`  Container:      ${indexConfig.containerName}`);
    console.log(`  Index type:     ${indexConfig.type}`);
    console.log(`  Dimensions:     ${config.expectedDimensions}`);
    console.log(`  Distance func:  cosine (queried with all 3 metrics)`);

    // Delete existing container to ensure clean state (idempotent)
    await deleteContainerIfExists(armClient, config, indexConfig.containerName);

    const start = Date.now();
    await armClient.sqlResources.beginCreateUpdateSqlContainerAndWait(
      config.azure.resourceGroup!,
      config.cosmos.accountName!,
      config.cosmos.databaseName,
      indexConfig.containerName,
      {
        resource: {
          id: indexConfig.containerName,
          partitionKey: {
            paths: ["/Region"],
            kind: "MultiHash",
            version: 2,
          },
          indexingPolicy: {
            indexingMode: "consistent",
            automatic: true,
            includedPaths: [{ path: "/*" }],
            excludedPaths: [{ path: "/_etag/?" }, { path: `${embeddingPath}/*` }],
            vectorIndexes: [
              {
                path: embeddingPath,
                type: indexConfig.type,
              },
            ],
          },
          vectorEmbeddingPolicy: {
            vectorEmbeddings: [
              {
                path: embeddingPath,
                dataType: "float32",
                dimensions: config.expectedDimensions,
                distanceFunction: "cosine",
              },
            ],
          },
        },
        location: config.azure.location,
      }
    );

    const elapsed = ((Date.now() - start) / 1000).toFixed(1);
    console.log(`  Created in ${elapsed}s`);
    console.log(
      "  Vector index is IMMUTABLE — cannot be changed after creation"
    );
  }
}

// ---------------------------------------------------------------------------
// Step 6 — Clean up sample-created containers (not azd infra)
// ---------------------------------------------------------------------------
export async function cleanupSampleContainers(
  armClient: CosmosDBManagementClient,
  config: SampleConfig
) {
  const containerNames = ["hotels_diskann_ts", "hotels_quantizedflat_ts"];

  console.log("\n=== Cleanup: Remove Sample Containers ===");
  for (const containerName of containerNames) {
    try {
      await armClient.sqlResources.beginDeleteSqlContainerAndWait(
        config.azure.resourceGroup!,
        config.cosmos.accountName!,
        config.cosmos.databaseName,
        containerName
      );
      console.log(`  ✓ Deleted container: ${containerName}`);
    } catch (error) {
      if ((error as any).code === 404 || (error as any).message?.includes("NotFound")) {
        console.log(`  ✓ Container does not exist: ${containerName}`);
      } else {
        throw error;
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Exports for testing
// ---------------------------------------------------------------------------
export { CosmosDBManagementClient } from "@azure/arm-cosmosdb";

File: nosql-create-index-dotnet/src/ControlPlane.cs
using Azure;
using Azure.Core;
using Azure.Identity;
using Azure.ResourceManager;
using Azure.ResourceManager.CosmosDB;
using Azure.ResourceManager.CosmosDB.Models;

namespace NosqlCreateIndexDotnet;

/// <summary>
/// Control-plane operations using Azure.ResourceManager.CosmosDB (ARM SDK).
/// 
///   1. Create containers with vector indexes (hotels_diskann_dotnet, hotels_quantizedflat_dotnet)
///   2. Clean up sample-created containers
/// 
/// RBAC Setup:
/// - Role definitions and assignments are created by `azd up` via Bicep
/// - Sample code uses DefaultAzureCredential() for authentication
/// - No RBAC creation code is needed in the sample
/// </summary>
public static class ControlPlane
{
    private const string PartitionKeyPath = "/Region";

    public static async Task CreateContainersAsync(
        SampleConfig config,
        TokenCredential credential,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(config);
        ArgumentNullException.ThrowIfNull(credential);

        var armClient = new ArmClient(credential);
        var accountIdentifier = CosmosDBAccountResource.CreateResourceIdentifier(
            config.SubscriptionId,
            config.ResourceGroup,
            config.AccountName);

        var accountResource = armClient.GetCosmosDBAccountResource(accountIdentifier);
        var account = await accountResource.GetAsync(cancellationToken);
        var database = await account.Value.GetCosmosDBSqlDatabaseAsync(config.DatabaseName, cancellationToken);
        var containers = database.Value.GetCosmosDBSqlContainers();

        var embeddingPath = $"/{config.EmbeddingFieldName}";

        // Create separate containers for each index type
        var indexConfigs = new[]
        {
            (Name: config.DiskANNContainerName, IndexType: CosmosDBVectorIndexType.DiskAnn),
            (Name: config.QuantizedFlatContainerName, IndexType: CosmosDBVectorIndexType.QuantizedFlat)
        };

        foreach (var indexConfig in indexConfigs)
        {
            Console.WriteLine("\n=== Step 1: Create Container with Vector Index ===");
            Console.WriteLine($"  Container:      {indexConfig.Name}");
            Console.WriteLine($"  Index type:     {indexConfig.IndexType}");
            Console.WriteLine($"  Dimensions:     {config.ExpectedDimensions}");
            Console.WriteLine($"  Distance func:  cosine (queried with all 3 metrics)");

            await DeleteContainerIfExistsAsync(containers, indexConfig.Name, cancellationToken);

            var start = DateTime.UtcNow;
            var containerDefinition = BuildContainerDefinition(
                indexConfig.Name, 
                indexConfig.IndexType,
                embeddingPath,
                config.ExpectedDimensions, 
                account.Value.Data.Location);
            
            await containers.CreateOrUpdateAsync(WaitUntil.Completed, indexConfig.Name, containerDefinition, cancellationToken);

            var elapsed = (DateTime.UtcNow - start).TotalSeconds;
            Console.WriteLine($"  Created in {elapsed:F1}s");
            Console.WriteLine("  Vector index is IMMUTABLE — cannot be changed after creation");
        }
    }

    public static async Task CleanupContainersAsync(
        SampleConfig config,
        TokenCredential credential,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(config);
        ArgumentNullException.ThrowIfNull(credential);

        var armClient = new ArmClient(credential);
        var accountIdentifier = CosmosDBAccountResource.CreateResourceIdentifier(
            config.SubscriptionId,
            config.ResourceGroup,
            config.AccountName);

        var accountResource = armClient.GetCosmosDBAccountResource(accountIdentifier);
        var account = await accountResource.GetAsync(cancellationToken);
        var database = await account.Value.GetCosmosDBSqlDatabaseAsync(config.DatabaseName, cancellationToken);
        var containers = database.Value.GetCosmosDBSqlContainers();

        var containerNames = new[] { config.DiskANNContainerName, config.QuantizedFlatContainerName };

        Console.WriteLine("\n=== Cleanup: Remove Sample Containers ===");
        foreach (var containerName in containerNames)
        {
            try
            {
                var existingContainer = await containers.GetIfExistsAsync(containerName, cancellationToken);
                if (existingContainer.HasValue)
                {
                    await existingContainer.Value!.DeleteAsync(WaitUntil.Completed, cancellationToken);
                    Console.WriteLine($"  ✓ Deleted container: {containerName}");
                }
                else
                {
                    Console.WriteLine($"  ✓ Container does not exist: {containerName}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"  ⚠ Failed to delete {containerName}: {ex.Message}");
            }
        }
    }

    private static CosmosDBSqlContainerCreateOrUpdateContent BuildContainerDefinition(
        string containerName,
        CosmosDBVectorIndexType indexType,
        string embeddingPath,
        int dimensions,
        AzureLocation location)
    {
        var indexingPolicy = new CosmosDBIndexingPolicy
        {
            IsAutomatic = true,
            IndexingMode = CosmosDBIndexingMode.Consistent
        };
        indexingPolicy.IncludedPaths.Add(new CosmosDBIncludedPath { Path = "/*" });
        indexingPolicy.ExcludedPaths.Add(new CosmosDBExcludedPath { Path = "/_etag/?" });
        indexingPolicy.ExcludedPaths.Add(new CosmosDBExcludedPath { Path = $"{embeddingPath}/*" });
        indexingPolicy.VectorIndexes.Add(new CosmosDBVectorIndex(embeddingPath, indexType));

        var resource = new CosmosDBSqlContainerResourceInfo(containerName)
        {
            PartitionKey = new CosmosDBContainerPartitionKey
            {
                Kind = CosmosDBPartitionKind.MultiHash,
                Version = 2
            },
            IndexingPolicy = indexingPolicy
        };
        resource.PartitionKey.Paths.Add(PartitionKeyPath);
        resource.VectorEmbeddings.Add(
            new CosmosDBVectorEmbedding(
                embeddingPath,
                CosmosDBVectorDataType.Float32,
                VectorDistanceFunction.Cosine,
                dimensions));

        return new CosmosDBSqlContainerCreateOrUpdateContent(location, resource);
    }

    private static async Task DeleteContainerIfExistsAsync(
        CosmosDBSqlContainerCollection containers,
        string containerName,
        CancellationToken cancellationToken)
    {
        Console.WriteLine("  Deleting existing container if present...");
        var existingContainer = await containers.GetIfExistsAsync(containerName, cancellationToken);
        if (existingContainer.HasValue)
        {
            await existingContainer.Value!.DeleteAsync(WaitUntil.Completed, cancellationToken);
            Console.WriteLine("  Deleted existing container");
        }
        else
        {
            Console.WriteLine("  Container does not exist (OK)");
        }
    }
}
